### ISSUE TITLE
Support multiple LDAP servers in a auth source

### ISSUE THREAD

**Issue by silverwind** [go-gitea/gitea#6898]
I have a LDAP auth source which has multiple redundant servers, I think it would be useful if a LDAP auth source would allow to specify more than one server here:

<img width="379" src="https://user-images.githubusercontent.com/115237/57516581-49bad800-7315-11e9-9ce3-ea42d6f4bad8.png">

Maybe accept a comma-separated list. Servers should be tried in the order they are defined, or possibly randomly to even the load. All servers should be tried and auth should only fail if it fails on all servers.

**Comment by silverwind** [comment]
Still want to do this, just haven't gotten around to it yet.

**Comment by X0rg** [comment]
Hello,

Any update about this PR? That is a feature I am looking for.
In resilient environments, having at least two LDAP servers is common (in case one server is not working properly for example), so I would like to put 2 URLs for my LDAP source.

I see content of `modules/auth/ldap/ldap.go` file was moved to `services/auth/source/ldap/source_search.go`, so this PR must be updated.

**Comment by palto42** [comment]
I would also be very interested in this feature in order to support higher availability.

**Comment by LucienCLEIA** [comment]
4 years after, there's still no changes for an issue related to security (availability is a part of security for me) ????
Are you sure you are right on your "What is Gitea?" page mention :

"Gitea places a strong emphasis on security, offering features such as user permission management, access control lists, and more to ensure the security of code and data."

I think this issue would have a better priority to be resolved in less than 4 years, no ?
As many others, I would also be very interested in this feature.

Thank you !

**Comment by techknowlogick** [comment]
@luCL21 We merge >400 PRs a month, and have thousands of support requests a month through forum, issues reports, emails, chat messages, and more. So while we wish to get to all of the feature requests sometimes it takes longer to get to everything. Part of our long-term strategic roadmap includes improving the high-availability of the project, and this would certainly fit in there.

That being said, we always welcome non-maintainer contributions, so I'll put up a bounty to perhaps usher the progress on this ticket along.

/bounty $200

**Comment by abhishek818** [comment]
@techknowlogick  Can i get this assigned?

**Comment by techknowlogick** [comment]
@abhishek818 yup. If you wish to attempt it for the bounty please ensure you follow the steps from algora posted above.

**Comment by wxiaoguang** [comment]
The "steps" proposed by algora-pbc seem to be easy but not quite right.

https://github.com/go-gitea/gitea/pull/31649#issuecomment-2505555133

> It needs to completely refactor the existing LDAP config related code, the key points (in my mind) could be:
>
> 1. use full LDAP URLs for servers, eg: `ldap://server:port/`,  `ldaps://server:port/`, `ldap+starttls://server:port/`
> 1. keep backwards compatibility (eg: avoid changing unnecessary CLI arguments, could re-use/migrate existing LDAP server configs in database)
